### PR TITLE
chore: no pointers to maps or slices in expconf

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

When I first designed the golang framework for expconf, I didn't actually realize that maps and arrays could be `nil`, so I was using redundant pointers to those objects.  This PR removes that.

## Test Plan

Added relevant objects to the test structs and tests still pass.

## Commentary

The only part of this change that I think is a bit opaque is the hardcoded whitelist of type aliases in gen.py.  To help improve the discoverability of the whitelist, I wrote very clear error messages when the dev might need to add an element to it.

### CODE REVIEW

File: master/pkg/schemas/defaults.go [hunk 1]
```diff
@@ -45,52 +45,55 @@ func getDefaultSource(obj reflect.Value) interface{} {
 	return nil
 }
 
+func allocateWithDefaultBytes(typ reflect.Type, defaultBytes []byte) reflect.Value {
```
Comment (stoksc) [inline master/pkg/schemas/defaults.go:48]: I'm beginning to think this should all return errors not panic. Yes, a panic means bad code, but that means bad code in a bad spot (not in an actor or http handler, which is basically none of our code but still) is unrecoverable which kinda freaks me out.

File: master/pkg/schemas/expconf/experiment_config.go [hunk 2]
```diff
@@ -117,6 +121,9 @@ func (l LabelsV0) MarshalJSON() ([]byte, error) {
 
 // UnmarshalJSON implements the json.Unmarshaler interface.
 func (l *LabelsV0) UnmarshalJSON(data []byte) error {
```
Comment (stoksc) [inline master/pkg/schemas/expconf/experiment_config.go:123]: remind me, why did/do we support a list and map here?

File: harness/determined/common/schemas/expconf/_gen.py [hunk 3]
```diff
@@ -2675,6 +2675,26 @@
                 "null"
             ],
             "default": null
+        },
```
Comment (stoksc) [inline harness/determined/common/schemas/expconf/_gen.py:2678]: why do we ever need to differentiate these? sorry coming back to this after not looking at it for a while.

**Review by stoksc — COMMENTED** [review]
lgtm, just some q's first to help me understand

File: master/pkg/schemas/expconf/experiment_config.go [hunk 4]
```diff
@@ -107,6 +108,9 @@ type LabelsV0 map[string]bool
 
 // MarshalJSON implements the json.Marshaler interface.
 func (l LabelsV0) MarshalJSON() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/pkg/schemas/expconf/experiment_config.go:110]: No; this is converting a set (`map[string]bool`) into a list.

[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/schemas/expconf/experiment_config.go:110]: oh yeh, i see now.

File: master/pkg/schemas/defaults.go (hunk 1 above)
Comment (rb-determined-ai) [reply master/pkg/schemas/defaults.go:48]: I mean, you've been saying that for a while.

I still don't think so though.  These errors can only occur with invalid default values, which are hardcoded json values and should be tested beforehand to be valid.

A better language would execute them at compile time as some sort of const expression.  In this case we can compensate for go with a bunch of tests.

But overall, something that conceptually means "take this struct and fill nil pointers with constant values" should not throw an error for anything other than a malloc failure (in which case Go panics anyway).

I could write a test that checks every default to make sure it is valid json if you'd like?

File: master/pkg/schemas/expconf/experiment_config.go (hunk 2 above)
Comment (rb-determined-ai) [reply master/pkg/schemas/expconf/experiment_config.go:123]: I have no idea.  I know it happened on commit `b8c5cdb72b` but I can't figure out why that was what we did.

File: harness/determined/common/schemas/expconf/_gen.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/common/schemas/expconf/_gen.py:2678]: Honoring `nil` values is equivalent to knowing if the user provided a value or not.  Trampling them would mean that you couldn't tell if you were looking at a defaulted value of if the user provided e.g. an empty list explicitly.

File: master/pkg/schemas/defaults.go (hunk 1 above)
Comment (stoksc) [reply master/pkg/schemas/defaults.go:48]: Nah that's way too much work. We should just be really careful to write tests for every expconf thing?

File: harness/determined/common/schemas/expconf/_gen.py (hunk 3 above)
Comment (stoksc) [reply harness/determined/common/schemas/expconf/_gen.py:2678]: Makes sense.

File: master/pkg/schemas/expconf/experiment_config.go (hunk 2 above)
Comment (stoksc) [reply master/pkg/schemas/expconf/experiment_config.go:123]: Weird. I wonder why it wasn't _always_ a list.

**Review by stoksc — APPROVED** [review]

**Pull request merged**